packages/chat-stateful-client/src/ChatClientState.ts

```diff
@@ -45,3 +51,82 @@ export type ChatThreadClientState = {
 export type ChatThreadProperties = {
   topic?: string;
 };
+
+/**
+ * Errors teed from API calls to the Chat SDK.
+ *
+ * Each property in the object stores the latest error for a particular SDK API method.
+ *
+ * Errors from this object can be cleared by calling the TODO(implement me) {@Link clearError} method.
+ * Additionally, errors are automatically cleared when:
+ * - The state is cleared.
+ * - Subsequent calls to related API methods succeed.
+ *   See documentation of individual stateful client methods for details on when errors may be automatically cleared.
+ */
+export type ChatErrors = {
+  [target in ChatErrorTargets]: Error;
+};
+
+/**
+ * Error thrown from failed stateful API methods.
+ */
+export class ChatError extends Error {
+  /**
+   * The API method target that failed.
+   */
+  public target: ChatErrorTargets;
+  /**
+   * Error thrown by the failed SDK method.
+   */
+  public inner: Error;
+
+  constructor(target: ChatErrorTargets, inner: Error) {
+    super();
+    this.target = target;
+    this.inner = inner;
+    this.name = 'ChatError';
+    this.message = `${this.target}: ${this.inner.message}`;
+  }
+}
+
+/**
+ * String literal type for all permissible keys in {@Link ChatErrors}.
+ */
+export type ChatErrorTargets =
+  | ChatObjectMethodNames<'ChatClient', ChatClient>
+  | ChatObjectMethodNames<'ChatThreadClient', ChatThreadClient>;
+
+/**
+ * Helper type to build a string literal type containing methods of an object.
+ */
+export type ChatObjectMethodNames<TName extends string, T> = {
+  [K in keyof T]: `${TName}.${ChatMethodName<T, K>}`;
+}[keyof T];
+
+/**
+ * Helper type to build a string literal type containing methods of an object.
+ */
+// eslint complains on all uses of `Function`. Using it as a type constraint is legitimate.
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type ChatMethodName<T, K extends keyof T> = T[K] extends Function ? (K extends string ? K : never) : never;
+
+/**
+ * Method to decide at runtime if a string is an error target.
+ */
+export const isChatErrorTarget = (target: string): boolean => {
+  for (const targetPrefix in ChatErrorTargetPrefixes) {
+    const [prefix, obj] = targetPrefix;
+    if (target.startsWith(prefix)) {
+      target = target.substring(prefix.length);
```

Comment:
nit/question: if target starts with prefix, and we take a substring of length prefix, does target == prefix? (i.e. could be `target = prefix;`)

Reply:
`target.substring(prefix.length)` gives back the substring of `target` *starting at* index `prefix.length` to the end of the string: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring

This should be covered by unit tests I added, so I'm confident this works as expected.